src/aap_eda/analytics/utils.py

```diff
@@ -141,3 +158,185 @@ def extract_job_details(
+
+    def is_expired(self) -> bool:
+        return int(time.time()) >= self.expires_at
+
+
+def generate_token() -> ServiceToken:
+    token_url = get_oidc_token_url()
+    client_id = get_client_id()
+    client_secret = get_client_secret()
+
+    if not all([client_id, client_secret, token_url]):
+        raise ValueError("Missing required OIDC configuration")
+
+    headers = {"Content-Type": "application/x-www-form-urlencoded"}
+
+    data = {
+        "client_id": client_id,
+        "client_secret": client_secret,
+        "grant_type": "client_credentials",
+    }
+
+    try:
+        resp = requests.post(
+            token_url,
+            headers=headers,
+            data=data,
+            verify=get_cert_path(),
+            timeout=(31, 31),
+        )
+        resp.raise_for_status()
+    except requests.exceptions.SSLError:
+        logger.error("SSL certificate verification failed")
+        raise
+    except requests.exceptions.Timeout:
+        logger.error("Token request timed out")
+        raise
+    except requests.exceptions.RequestException as e:
+        logger.error(f"Token request failed: {str(e)}")
+        raise
+
+    data = resp.json()
+
+    return ServiceToken(
+        access_token=data.get("access_token", ""),
+        expires_in=data.get("expires_in", 0),
+    )
+
+
+def _get_credential_value(field: str, *setting_envs: Tuple[Any, str]) -> str:
+    credentials = _get_analytics_credentials()
+    for credential in credentials:
+        inputs = inputs_from_store(credential.inputs.get_secret_value())
+        if value := inputs.get(field):
+            return value
+
+    for env, key in setting_envs:
+        if value := getattr(env, key, None):
+            return value
+
+
+def get_cert_path() -> str:
+    return settings.INSIGHTS_CERT_PATH
+
+
+def get_auth_mode() -> str:
+    return settings.AUTOMATION_AUTH_METHOD
+
+
+def get_oidc_token_url() -> str:
+    return settings.OIDC_TOKEN_URL
+
+
+def get_proxy_url() -> str:
+    return settings.ANALYTICS_PROXY_URL
+
+
+def get_analytics_url() -> str:
+    return (
+        application_settings.AUTOMATION_ANALYTICS_URL
+        or settings.AUTOMATION_ANALYTICS_URL
+    )
+
+
+def get_client_id() -> str:
+    return _get_credential_value("client_id")
+
+
+def get_client_secret() -> str:
+    return _get_credential_value("client_secret")
+
+
+def get_username() -> str:
+    _validate_credential()
+    return _get_credential_value(
+        "username",
+        (settings, "REDHAT_USERNAME"),
+        (application_settings, "REDHAT_USERNAME"),
+        (application_settings, "SUBSCRIPTIONS_USERNAME"),
+    )
+
+
+def get_password() -> str:
+    _validate_credential()
+    return _get_credential_value(
+        "password",
+        (settings, "REDHAT_PASSWORD"),
+        (application_settings, "REDHAT_PASSWORD"),
+        (application_settings, "SUBSCRIPTIONS_PASSWORD"),
+    )
+
+
+def get_insights_tracking_state() -> str:
```

Comment:
@hsong-rh since we are returning a boolean or a string from this function should the return type be fixed.